hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/io/storage/HoodieSparkIOFactory.java

```diff
@@ -51,4 +66,30 @@ public HoodieFileWriterFactory getWriterFactory(HoodieRecord.HoodieRecordType re
     }
     return super.getWriterFactory(recordType);
   }
+
+  @Override
+  public Function<HoodieRecord<?>, IndexedRecord> toIndexedRecord(Schema recordSchema, Properties properties, HoodieRecord.HoodieRecordType recordType) {
```

Comment:
I'm wondering if we can further separate the read and write paths that use the merge API to use different merger for now.  For the write path that still uses writer expecting Avro records (e.g., Avro parquet writer), we still use Avro-based merger; for read path that expects engine-record, i.e., InternalRow/Row in Spark, the Internal Row should be used for merging.  In Hudi 1.1, we can reduce the complexity.

Reply:
Right now, the writer side has gaps on fully supporting Spark merger.

Reply:
ok got rid of this. Pulled the changes into a new pr: https://github.com/apache/hudi/pull/12115

Reply:
It seems like these changes are still here.